cirq/optimizers/cphase_to_fsim.py

```diff
@@ -0,0 +1,218 @@
+
+    Raises:
+        ValueError under any of the following circumstances:
+         * cphase_gate or fsim_gate is parametrized,
+         * cphase_gate and fsim_gate do not satisfy the conditions above,
+         * fsim_gate has invalid angles (see atol argument above),
+         * incorrect number of qubits are provided.
+    """
+    if protocols.is_parameterized(cphase_gate):
+        raise ValueError('Cannot decompose a parametrized gate.')
+    if protocols.is_parameterized(fsim_gate):
+        raise ValueError('Cannot decompose into a parametrized gate.')
+    if qubits is None:
+        qubits = devices.LineQubit.range(2)
+    if len(qubits) != 2:
+        raise ValueError(f'Expected a pair of qubits, but got {qubits!r}.')
+    q0, q1 = qubits
+
+    theta = fsim_gate.theta
+    phi = fsim_gate.phi
+
+    sin_half_phi = np.sin(phi / 2)
+    cos_half_phi = np.cos(phi / 2)
+    sin_theta = np.sin(theta)
+    cos_theta = np.cos(theta)
+
+    #
+    # Step 1: find alpha
+    #
+    denominator = (sin_theta - sin_half_phi) * (sin_theta + sin_half_phi)
+    if abs(denominator) < atol:
+        raise ValueError(
+            f'{fsim_gate} cannot be used to decompose CZPowGate because '
+            'sin(theta)**2 is too close to sin(phi/2)**2 '
+            f'(difference is {denominator}).')
+
+    # Parametrization of CZPowGate by a real angle is a non-injective function
+    # with the preimage of cphase_gate infinite. However, it is sufficient to
+    # check just two of the angles against the constraints of the decomposition.
+    canonical_delta = -np.pi * (cphase_gate.exponent % 2)
+    for delta in (canonical_delta, canonical_delta + 2 * np.pi):
+        sin_quarter_delta = np.sin(delta / 4)
+        numerator = (sin_quarter_delta - sin_half_phi) * (sin_quarter_delta +
+                                                          sin_half_phi)
+        sin_alpha_squared = numerator / denominator
+        if 0 <= sin_alpha_squared <= 1:
+            break
+    else:
+        intervals = compute_cphase_exponents_for_fsim_decomposition(fsim_gate)
+        raise ValueError(
+            f'{cphase_gate} cannot be decomposed into two {fsim_gate}. Valid '
+            f'intervals for canonical exponent of CZPowGate: {intervals}.')
+    assert 0 <= sin_alpha_squared <= 1
+    alpha = np.arcsin(np.sqrt(sin_alpha_squared))
+
+    #
+    # Step 2: find xi and eta
+    #
+    tan_alpha = np.tan(alpha)
+    xi = np.arctan2(tan_alpha * cos_theta, cos_half_phi)
+    eta = np.arctan2(tan_alpha * sin_theta, sin_half_phi)
+    if delta < 0:
+        eta += np.pi
+
+    #
+    # Step 3: synthesize output circuit
+    #
+    return circuits.Circuit(
+        # Local X rotations to convert Γ1⊗I − iZ⊗Γ2 into exp(-i Z⊗Z δ/4)
+        ops.rx(xi).on(q0),
+        ops.rx(eta).on(q1),
+
+        # Y(θ, φ) := exp(-i X⊗X θ/2) exp(-i Y⊗Y θ/2) exp(-i Z⊗Z φ/4)
+        fsim_gate.on(q0, q1),
+        ops.rz(phi / 2).on(q0),
+        ops.rz(phi / 2).on(q1),
```

Comment:
Consider using `PhasedXZ` instead of separate `Rx` and `Rz` gates.

Reply:
I like the transparency and readability of the operations here and I'm concerned that we would loose it (as a result of some gates straddling the conceptually distinct steps in the decomposition).

Perhaps we can leave such optimization up to the caller (and another optimizer) as we did with merging single-qubit gates and dropping empty moments in Wojtek's comment earlier?

Reply:
I agree with Adam here.